internal/collection/serde.go

```diff
@@ -0,0 +1,855 @@
+		log.Errorf("Failed to write snapshotId into tree data file")
+		return
+	}
+
+	if err = binary.Write(bdw, binary.BigEndian, uint32(c.items.Len())); err != nil {
+		log.Errorf("Failed to write tree length into items data file")
+		return
+	}
+
+	itemMap = make(map[*itemT]uint32, c.items.Len())
+	var itemNum uint32
+	itemSaver := func (w io.Writer, value interface{}) (err error) {
+		item := value.(*itemT)
+		itemMap[item] = itemNum
+		if err = binary.Write(w, binary.BigEndian, itemNum); err != nil {
+			return
+		}
+
+		// using closure to access data file, into which we dump the actual data
+		if err = saveString(bdw, item.id); err != nil {
+			return
+		}
+		if err = binary.Write(bdw, binary.BigEndian, int32(item.fieldValuesSlot)); err != nil {
+			return
+		}
+		if err = binary.Write(bdw, binary.BigEndian, objIsSpatial(item.obj)); err != nil {
+			return
+		}
+		if err = saveString(bdw, item.obj.String()); err != nil {
+			return
+		}
+
+		itemNum++
+		return
+	}
+
+	if err = c.items.Save(btw, itemSaver); err != nil {
+		log.Errorf("Failed to save items tree and data")
+	}
+
+	if err = binary.Write(bdw, binary.BigEndian, snapshotId); err != nil {
+		log.Errorf( "Failed to write snapshotId into items data file")
+		return
+	}
+
+	if err = binary.Write(btw, binary.BigEndian, snapshotId); err != nil {
+		log.Errorf("Failed to write snapshotId into items tree file")
+		return
+	}
+
+	return
+}
+
+func (c * Collection) loadItemsData(dataFile string, snapshotId uint64, parseOpts *geojson.ParseOptions) (itemList []*itemT, err error) {
+	var f *os.File
+	f, err = os.Open(dataFile)
+	log.Infof("Opened itemsData file: %s", dataFile)
+	if err != nil {
+		return
+	}
+	defer func() {
+		if f.Close() != nil {
+			log.Errorf("Failed to close %s", dataFile)
+		}
+	}()
+	br := bufio.NewReader(f)
+
+	if err = verifySnapshotId(br, snapshotId); err != nil {
+		return
+	}
+
+	var word uint32
+	if err = binary.Read(br, binary.BigEndian, &word); err != nil {
+		log.Errorf("Failed to read tree length from items data file")
+	}
+	nItems := int(word)
+	itemList = make([]*itemT, nItems)
+	buf := make([]byte, 0)
+	var idStr, objStr string
+	var spatial bool
+	var fvs int32
+
+	type todoItem struct {
+		index int
+		spatial bool
+		id string
+		json string
+		slot fieldValuesSlot
+	}
+
+	todoChannel := make(chan todoItem, 1024)
+
+	var wg sync.WaitGroup
+	var nWorkers int
+	if runtime.NumCPU() > 10 {
+		nWorkers = 10
+	} else {
+		nWorkers = runtime.NumCPU()
+	}
+	for i := 0; i < nWorkers; i++ {
+		go func() {
+			defer wg.Done()
+			var obj geojson.Object
+			for ti := range todoChannel {
+				if ti.spatial {
+					if obj, err = geojson.Parse(ti.json, parseOpts); err != nil {
+						log.Errorf("Failed to parse object from data file, json %s", ti.json)
+						return
+					}
+				} else {
+					obj = String(ti.json)
+				}
+				itemList[ti.index] = &itemT{id: ti.id, obj: obj, fieldValuesSlot: ti.slot}
+			}
+		}()
+		wg.Add(1)
```

Comment:
The add should happen prior to launching the goroutine

Reply:
Done